### PR TITLE
[IO-461][external] Convert RG16 images into grayscale

### PR DESCRIPTION
**Pull request by rslota**
See: https://linear.app/v7labs/issue/IO-461/request-darwin-dataset-pull-command-with-video-frames-option-downloads

### CODE REVIEW

File: darwin/dataset/download_manager.py [hunk 1]
```diff
@@ -464,14 +473,33 @@ def _fetch_multiple_files(path: Path, response: requests.Response) -> None:
...
-def _write_file(path: Path, response: requests.Response) -> None:
+def _write_file(path: Path, response: requests.Response, transform_file_function=None) -> None:
     with open(str(path), "wb") as file:
         for chunk in response:
             file.write(chunk)
+    if transform_file_function is not None:
+        transform_file_function(path)
+
+
+def _rg16_to_grayscale(path):
```
Comment (tomvars) [inline darwin/dataset/download_manager.py:491]: feel like this helper function should live in a `utils.py` style file not here

**Review by tomvars — CHANGES_REQUESTED** [review]
Looks good to me, haven't tested it but let me know if you need me to.

File: darwin/dataset/download_manager.py (hunk 1 above)
Comment (rslota) [reply darwin/dataset/download_manager.py:491]: Why though? It's only used here. If we ever need it somewhere else, we could move it and make it more public then.

**Comment by rslota** [comment]
> Looks good to me, haven't tested it but let me know if you need me to.

I've tested this with DICOM file I found in busybee tests which is RG16 and one of mine that results in RGB. I tested this for both V1 and V2 with both Darwin JSON 1.0 and Darwin JSON 2.0.

File: darwin/dataset/download_manager.py (hunk 1 above)
Comment (tomvars) [reply darwin/dataset/download_manager.py:491]: ok, agreed its unlikely to be used elsewhere 👍

**Review by tomvars — APPROVED** [review]

File: darwin/dataset/download_manager.py [hunk 2]
```diff
@@ -464,14 +473,33 @@ def _fetch_multiple_files(path: Path, response: requests.Response) -> None:
...
+    image_2d_rgb = np.asarray(image)
+
+    image_2d_r = np.uint16(image_2d_rgb[:, :, 0]) << 8
+    image_2d_g = np.uint16(image_2d_rgb[:, :, 1])
+
+    image_2d_gray = np.bitwise_or(image_2d_r, image_2d_g)
+    image_2d_gray = image_2d_gray / (1 << 16) * 255
+
+    new_image = Image.fromarray(np.uint8(image_2d_gray), mode="L")
+    new_image.save(path)
```
Comment (owencjones) [inline darwin/dataset/download_manager.py:505]: This is doing a decent job of the RGB -> Grayscale conversion based on focussing on R and G is better than the mean luminescence of R,G, and B, but if you find the outputs are lacking, the coeffiecients below are useful for grayscale transformation:

If so, here they are:
```
grey = 0.2989 * R + 0.5870 * G + 0.1140 * B
```

Comment (rslota) [reply darwin/dataset/download_manager.py:505]: This is not RGB input though. This is our custom format that encodes 16 grayscale image on first 2 bytes of RGB tuple. "B" channel is all zeros in that case. Those files are generated by our backend here: https://github.com/v7labs/busybee/blob/master/dicom-extractor/dicom-extractor.py#L72

Comment (owencjones) [reply darwin/dataset/download_manager.py:505]: Ah right, didn't realise this was a DICOM input.   Yeah, in that case my earlier comment is not at all applicable!

**Pull request merged**